[HLSL][docs] Document hlsl.h in the HLSL docs

llvm-beanz opened the pull request:
This adds a brief blurb about hlsl.h in the HLSLSupport documentation where a high level view of the architecture is explained.

llvm-beanz committed:
[HLSL][docs] Document hlsl.h in the HLSL docs

This adds a brief blurb about hlsl.h in the HLSLSupport documentation
where a high level view of the architecture is explained.

bharadwajy reviewed (approved):
Consider consistent spelling of the term "built in" consistently - whichever is preferred.

llvm-beanz committed:
Clarify use and spelling of the word "builtin"

farzonl commented on clang/docs/HLSL/HLSLSupport.rst:
```diff
@@ -114,6 +114,44 @@ not re-targetable, we want to share the Clang CodeGen implementation for HLSL
+distributed as part of the Clang resource directory.
+
+Similar to OpenCL, HLSL's implicit header will explicitly declare all overloads,
+and each overload will map to a corresponding ``__builtin_*`` compiler intrinsic
```
__builtin_* --> __builtin* 

seems to be how you do it elsewhere.

farzonl commented:
Should we talk at all about why we use __attribute__ clang_builtin_alias similar to what risc-v vectors does in their docs? https://github.com/llvm/llvm-project/blob/55c466da2f2f0baa509eb709b8de8926bd498b9b/clang/include/clang/Basic/AttrDocs.td#L5425

bharadwajy commented on clang/docs/HLSL/HLSLSupport.rst:
```diff
@@ -114,6 +114,44 @@ not re-targetable, we want to share the Clang CodeGen implementation for HLSL
+where parsing is a significant compile-time overhead, but the HLSL implicit
+headers can be compiled into a module for performance if needed.
+
+Further, by treating these functions as functions rather than ``__builtin*``
```
```suggestion
Further, by treating these as traditional language functions rather than as ``__builtin*`` 
```

bharadwajy commented on clang/docs/HLSL/HLSLSupport.rst:
```diff
@@ -114,6 +114,44 @@ not re-targetable, we want to share the Clang CodeGen implementation for HLSL
+hlsl.h
+------
+
+HLSL has an extensive library of functionality. This is similar to OpenCL and
```
```suggestion
HLSL has a library that supports extensive set of functionalities. This is similar to OpenCL and
```

bharadwajy commented on clang/docs/HLSL/HLSLSupport.rst:
```diff
@@ -114,6 +114,44 @@ not re-targetable, we want to share the Clang CodeGen implementation for HLSL
+headers can be compiled into a module for performance if needed.
+
+Further, by treating these functions as functions rather than ``__builtin*``
+compiler intrinsics the language behaviors are more consistent and aligned with
```
```suggestion
compiler intrinsics, the language behaviors are more consistent and aligned with
```

python3kgae commented on clang/docs/HLSL/HLSLSupport.rst:
```diff
@@ -114,6 +114,44 @@ not re-targetable, we want to share the Clang CodeGen implementation for HLSL
+and each overload will map to a corresponding ``__builtin_*`` compiler intrinsic
+that is handled in ClangCodeGen. CUDA uses a similar pattern although many CUDA
+functions have full definitions in the included headers which in turn call
+corresponding ``__builtin*`` compiler intrinsics. By not having bodies HLSL
```
Is it possible to define the body for some hlsl intrinsic functions like https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-d3dcolortoubyte4 which doesn't map to either DXIL or SPIRV?

llvm-beanz replied:
Possible, sure, but there is a performance impact do doing that because the function will need to be parsed, generated and inlined.

damyanp commented on clang/docs/HLSL/HLSLSupport.rst:
```diff
@@ -114,6 +114,44 @@ not re-targetable, we want to share the Clang CodeGen implementation for HLSL
+
+HLSL has an extensive library of functionality. This is similar to OpenCL and
+CUDA. The implementation approach for the HLSL library functionality draws from
+patterns in use by OpenCL and other Clang resource headers.
```
Would it be possible to have links to these similar headers for reference?

llvm-beanz committed:
Updates based on review feedback.

Also added a note about complex data types being defined in the
`HLSLExternalSemaSource`.

llvm-beanz commented on clang/docs/HLSL/HLSLSupport.rst:
```diff
@@ -114,6 +114,54 @@ not re-targetable, we want to share the Clang CodeGen implementation for HLSL
+``ClangHeaders`` component found in the source tree under
+`clang/lib/Headers <https://github.com/llvm/llvm-project/tree/main/clang/lib/Headers>`_.
+
+.. note::
```
Comment for reviewers: GitHub's preview doesn't properly render the `note` block. An example of this in HTML can be found here: https://clang.llvm.org/docs/LanguageExtensions.html#has-builtin
